File: apps/api/app/models/signal.py
"""
交易信号模型
"""

from datetime import datetime
from decimal import Decimal
from typing import Optional

from sqlalchemy import (
    String,
    Numeric,
    Index,
    ForeignKey,
    JSON,
    Boolean,
    Text,
)
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.db.base import Base


class Signal(Base):
    """交易信号记录表"""

    __tablename__ = "signals"
    __table_args__ = (
        Index("idx_signal_code", "code", "created_at"),
        Index("idx_signal_type", "signal_type", "created_at"),
        Index("idx_signal_date", "created_at"),
    )

    id: Mapped[int] = mapped_column(primary_key=True, autoincrement=True)
    created_at: Mapped[datetime] = mapped_column(default=datetime.utcnow, comment="信号生成时间")
    code: Mapped[str] = mapped_column(
        ForeignKey("stocks.code", ondelete="CASCADE"), nullable=False, comment="股票代码"
    )
    signal_type: Mapped[str] = mapped_column(
        String(20), nullable=False, comment="信号类型: buy/sell/watch/alert"
    )
    confidence: Mapped[Decimal] = mapped_column(
        Numeric(3, 2), nullable=False, comment="置信度 0.00-1.00"
    )
    trigger_price: Mapped[Optional[Decimal]] = mapped_column(
        Numeric(10, 4), nullable=True, comment="触发价格"
    )
    reason: Mapped[Optional[str]] = mapped_column(Text, nullable=True, comment="信号原因描述")
    agents_votes: Mapped[Optional[dict]] = mapped_column(
        JSON, nullable=True, comment="各Agent投票详情"
    )
    extra_data: Mapped[Optional[dict]] = mapped_column(
        JSON, nullable=True, comment="额外元数据"
    )
    is_valid: Mapped[bool] = mapped_column(default=True, comment="是否有效（复盘时标记）")

    # 关联关系
    stock: Mapped["Stock"] = relationship("Stock", back_populates="signals")
    decisions: Mapped[list["AgentDecision"]] = relationship(
        "AgentDecision", back_populates="signal"
    )

File: apps/api/internal/model/signal.go
package model

import (
	"time"

	"gorm.io/gorm"
)

// Signal 交易信号记录表
type Signal struct {
	ID        uint      `json:"id" gorm:"primaryKey"`
	CreatedAt time.Time `json:"created_at"`

	Code       string  `json:"code" gorm:"index;size:20;not null"`
	SignalType string  `json:"signal_type" gorm:"size:20;not null"` // buy/sell/watch/alert
	Confidence float64 `json:"confidence" gorm:"type:decimal(3,2);not null"` // 0.00-1.00
	TriggerPrice *float64 `json:"trigger_price,omitempty" gorm:"type:decimal(10,4)"`
	Reason     string  `json:"reason,omitempty" gorm:"type:text"`
	AgentsVotes JSON    `json:"agents_votes,omitempty" gorm:"type:jsonb"`
	ExtraData  JSON    `json:"extra_data,omitempty" gorm:"type:jsonb"`
	IsValid    bool    `json:"is_valid" gorm:"default:true"` // 复盘时标记

	// 关联关系
	Stock     Stock          `json:"stock,omitempty" gorm:"foreignKey:Code;references:Code"`
	Decisions []AgentDecision `json:"decisions,omitempty" gorm:"foreignKey:SignalID"`
}

// TableName 指定表名
func (Signal) TableName() string {
	return "signals"
}

// BeforeCreate 创建前钩子
func (s *Signal) BeforeCreate(tx *gorm.DB) error {
	if s.CreatedAt.IsZero() {
		s.CreatedAt = time.Now()
	}
	return nil
}

// AgentDecision Agent决策详情表
type AgentDecision struct {
	ID        uint      `json:"id" gorm:"primaryKey"`
	CreatedAt time.Time `json:"created_at"`

	SignalID  uint    `json:"signal_id" gorm:"index;not null"`
	AgentType string  `json:"agent_type" gorm:"size:30;not null"`
	Decision  string  `json:"decision" gorm:"size:10;not null"` // buy/sell/hold
	Score     *float64 `json:"score,omitempty" gorm:"type:decimal(3,2)"`
	Reasoning string  `json:"reasoning,omitempty" gorm:"type:text"`
	Weight    float64 `json:"weight" gorm:"type:decimal(3,2);default:1.0"`

	// 关联关系
	Signal Signal `json:"signal,omitempty" gorm:"foreignKey:SignalID"`
}

// TableName 指定表名
func (AgentDecision) TableName() string {
	return "agent_decisions"
}
